Fix SplitGraphs bug

wouterjsmit opened the pull request:
Fixes the bug of drawing too many lines for the subset of genomes.

See #169.

To fix this, I corrected the `getGenomesOverEdge` method. This method was rather an estimate than the exact genomes, which I did need however for the SplitGraphs algorithm to select the right edges.

Right now, the method uses IDs, but a more solid value should be used (such as `getLevel()`). This can be done as soon as the semantic branch has been merged.

A ticket for this can be found here: #186

wouterjsmit committed:
Fix inheritance mistake in SplitGraphs tests

The nodes in a graph are not necessarily SequenceNodes.

wouterjsmit committed:
Fix bug in node mocking in SplitGraphs

wouterjsmit committed:
Add test for edges in SplitGraphs

This test asserts that only the correct edges are drawn. (and that all
correct edges are drawn)

wouterjsmit committed:
Add test for copying nodes in SplitGraphs

Asserts that new nodes are created, rather than old ones reused.

wouterjsmit committed:
Fix bug in getGenomesOverEdge in GraphNode

It would return the edge over a genome even if that genome actually went
through another edge, via a longer path.

This is related to
https://github.com/ProgrammingLife2016/PL2-2016/issues/169

wouterjsmit committed:
Fix SplitGraphs bug that draws too many edges

Fixes: https://github.com/ProgrammingLife2016/PL2-2016/issues/169

justinvdk commented on PL2/PL2-core/src/main/java/nl/tudelft/pl2016gr2/core/algorithms/subgraph/SplitGraphs.java:
```diff
@@ -89,53 +91,55 @@ public SequenceGraph getSubgraph(Collection<Integer> genomes) {
+      if (node.getId() == 8720) {
+        System.out.println("Arrived at node 8720");
+        node.getOutEdges().forEach(System.out::println);
+      }
```
These lines seem be to for debugging purposes.

justinvdk commented:
I'm not qualified to talk about correctness, but the code seems to be in order.
I didn't see any odd edges in the graph either.

wouterjsmit commented:
Merging if no further comments?

FarisElghlan commented on PL2/PL2-core/src/main/java/nl/tudelft/pl2016gr2/core/algorithms/subgraph/SplitGraphs.java:
```diff
@@ -74,68 +73,63 @@ public SequenceGraph getSubgraph(Collection<Integer> genomes) {
+    newNodes.forEach((id, node) -> {
+      node.addAllInEdges(pruneInEdges(node, newNodes, oldNodes, genomes));
+      node.addAllOutEdges(pruneOutEdges(node, newNodes, oldNodes, genomes));
+      node.addAllGenomes(pruneGenomes(oldNodes.get(id), genomes));
```
Shouldn't `set...` be used here instead of `addAll` for all 3 of these?
I think this causes a bug for the genomes, since the copy method coppies the genomes and now you add the genomes to the coppied arraylist, resulting in an arraylist with many duplicates and some non-duplicates which shouldn't have been in the genome list at all.
Note however that the main graph is not altered, since the genome lists are coppied into new arraylist when the node is coppied. Maybe we could change this for some of the lists (if we are sure they aren't altered) so their references are passed (or at least for the `copyAll` methods), since copying all of the lists if they won't be altered is bad for perfomance and memory usage.

wouterjsmit replied:
I originally meant the `copy()` method to only copy the ID. If that is no longer the case however, you are right and that must be changed.

I removed the `set...()` because I think it is bad practice to allow the caller to specify the collection and its pointer (or the `set...()` methods should remove the old fields and add all of the specified Collection).

I agree that immutable fields would be really nice. I decided not pursue this because of compatibility and limited time. When everything is put together again, we could refactor this.

FarisElghlan replied:
ok apparently the implementation of the copy method on dev is different from the copy method on the branch which I am currently working on (zoom bubbles)

wouterjsmit replied:
The zoom bubbles could still adhere to the rule of not copying the genomes,inedges and outedges, as is specified by `GraphNode.copy()`.

I will add tests that assert this is the case to the SequenceNode test suite (which is now at 44 tests).
This suite should be refactored so that the GraphNode interface can be tested for all bubbles. Especially for tests for copying that is useful.

FarisElghlan commented:
Found one possible bug, the rest looks good.

wouterjsmit committed:
Add tests for GraphNode.copy()

These tests assert that the in edges, out edges and genomes are empty in
the copied instance.

wouterjsmit commented:
Added tests that verify absence of this bug. Note that the suite should be extended to work for all `GraphNode`s at some point soon.
